client-core/src/wallet/syncer.rs

```diff
@@ -0,0 +1,546 @@
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Fast forwards state to given status if app hashes match
+    fn fast_forward_status(&self, status: &Status) -> Result<Option<BlockHeader>> {
+        let current_app_hash = status
+            .sync_info
+            .latest_app_hash
+            .ok_or_else(|| Error::new(ErrorKind::TendermintRpcError, "latest_app_hash not found"))?
+            .to_string();
+
+        if current_app_hash == self.sync_state.last_app_hash {
+            let current_block_height = status.sync_info.latest_block_height.value();
+
+            let block = self.env.client.block(current_block_height)?;
+            let block_result = self.env.client.block_results(current_block_height)?;
+
+            Ok(Some(prepare_block_header(
+                &self.wallet,
+                &block,
+                &block_result,
+            )?))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Fast forwards state to given block if app hashes match
+    fn fast_forward_block(&mut self, block: &Block) -> Result<Option<BlockHeader>> {
+        let current_app_hash = block
+            .header
+            .app_hash
+            .err_kind(ErrorKind::TendermintRpcError, || "app_hash not found")?
+            .to_string();
+
+        if current_app_hash == self.sync_state.last_app_hash {
+            let current_block_height = block.header.height.value();
+            let block_result = self.env.client.block_results(current_block_height)?;
+            Ok(Some(prepare_block_header(
+                &self.wallet,
+                &block,
+                &block_result,
+            )?))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+fn decrypt_staking_transactions(
+    block_results: &BlockResults,
+    staking_addresses: &BTreeSet<StakedStateAddress>,
+    block: &Block,
+) -> Result<Vec<Transaction>> {
+    for staking_address in staking_addresses {
+        if block_results.contains_account(&staking_address)? {
+            return block.staking_transactions();
+        }
+    }
+
+    Ok(Default::default())
+}
+
+/// decrypt block for wallet
+fn prepare_block_header(
+    wallet: &Wallet,
+    block: &Block,
+    block_result: &BlockResults,
+) -> Result<BlockHeader> {
+    let app_hash = block
+        .header
+        .app_hash
+        .ok_or_else(|| Error::new(ErrorKind::TendermintRpcError, "app_hash not found"))?
+        .to_string();
+    let block_height = block.header.height.value();
+    let block_time = block.header.time;
+
+    let valid_transaction_ids = block_result.transaction_ids()?;
+    let block_filter = block_result.block_filter()?;
+
+    let staking_transactions =
+        decrypt_staking_transactions(&block_result, &wallet.staking_addresses(), block)?;
+
+    let enclave_transaction_ids = if block_filter.check_view_key(&wallet.view_key.clone().into()) {
+        block.enclave_transaction_ids()?
+    } else {
+        vec![]
```

Comment:
The logic depending on an empty vector may be a bit puzzling / indirect: one may expect that information in `BlockHeader` is kind of "raw" / wallet-independent (just processing what's returned from RPC), but this creates an extra hidden assumption around the view key.

Reply:
But `prepare_block_header` already filter staking transactions for current wallet, that's why I think it's ok to move this logic here.

Reply:
Maybe we should move `struct BlockHeader` into `syncer`, and make it private, there are no other code use it.

Reply:
And the `Header` part of the name is not accurate, because it actually contains transactions. So maybe change it into `FilteredBlock`?

Reply:
It's just an extra implicit invariant around `BlockHeader` (one could still construct BlockHeader that breaks this invariant). To make it more explicit, as you proposed, you can move to syncer, make it private and rename as `FilteredBlock`.

(The other option would be to move this prepare function into `impl FilteredBlock/BlockHeader ` and make it possible to only construct via this function, e.g. by making some of those fields private... but the first option may be simpler)

Reply:
Done both.